### PR TITLE
Expose FakeTimer, add FakeAsync.pendingTimers

### PR DESCRIPTION
**Pull request by jakemac53**
Ports https://github.com/google/quiver-dart/commit/0a282a82125b011152fab57185f9d39cfe71b12e to this package.

This is necessary to migrate `flutter_test` to use this package and drop its quiver dependency.

cc @jonahwilliams

### CODE REVIEW

File: lib/fake_async.dart [hunk 1]
```diff
@@ -272,13 +276,23 @@ class _FakeTimer implements Timer {
   /// fired.
   Duration _nextCall;
 
+  /// The current stack trace when this timer was created.
+  final StackTrace _creationStackTrace;
```
Comment (natebosch) [inline lib/fake_async.dart]: You could inline the initialization

final _creationStackTrace = StackTrace.current;

File: lib/fake_async.dart [hunk 2]
```diff
@@ -272,13 +276,23 @@ class _FakeTimer implements Timer {
...
+
   @override
   int get tick {
     throw UnimplementedError('tick');
   }
 
+  /// Returns debugging information to try to identify the source of the
```
Comment (natebosch) [inline lib/fake_async.dart:290]: can you expand this and say what information it includes?

File: lib/fake_async.dart [hunk 3]
```diff
@@ -68,6 +68,10 @@ class FakeAsync {
   /// All timers created within [run].
   final _timers = <_FakeTimer>{};
 
+  /// Debugging information for all pending timers.
+  List<String> get pendingTimersDebugInfo =>
+      _timers.map((timer) => '${timer.debugInfo}').toList(growable: false);
```
Comment (natebosch) [inline lib/fake_async.dart]: We don't need the extra `'${}'` here. `debugInfo` is already a String.

File: lib/fake_async.dart (hunk 1 above)
Comment (jakemac53) [reply lib/fake_async.dart]: Ah sure, I was just keeping the implementations identical but might as well.

File: lib/fake_async.dart (hunk 2 above)
Comment (jakemac53) [reply lib/fake_async.dart:290]: done

File: lib/fake_async.dart (hunk 3 above)
Comment (jakemac53) [reply lib/fake_async.dart]: Ya just copy/pasted from quiver, ill update though

**Review by natebosch — APPROVED** [review]
I think this would be easier to use programmatically if we returned an object with a field for the stack trace instead of a String, but it's probably also fine to have an API designed mostly around human readability that doesn't encourage doing something tricky with it in your tests...

If this is what flutter is hoping for I don't think we need to worry too much over the design.

**Comment by jakemac53** [comment]
@cbracken it actually looks like for flutter exposing the original information (and not a String) would be more useful, its just recreating the stack trace anyways and doing some [fragile munging of the message](https://github.com/flutter/flutter/blob/master/packages/flutter_test/lib/src/binding.dart#L1126).

If we changed this to return a `List<TimerDebugInfo>` or possibly even just exposed a `List<FakeTimer>` (and made FakeTimer public) would that inhibit quivers ability to re-export this package instead of its own implementation? Presumably there will be a breaking release for NNBD anyways.

**Comment by cbracken** [comment]
I've got an NNBD update to 3.0.0 in the works. Are you thinking of taking on a dependency on package:stack_trace, or authoring something minimal yourself?

Quiver has historically avoided almost all dependencies because frankly, I'm tired of migrating packages from one abandoned package to another every time the original author decides they have better things to do than maintain their code. A perfect example is the chain of 3-4 migrations that package:coverage had to go through dealing with the vm service, when its original implementation worked fine across every version of the VM service protocol between the original version and the latest and took up all of 100 lines of code without incurring 30 packages worth of transitive deps. </rant>

In fairness, I sort of assume package:stack_trace will be maintained. Though I suppose I'd have naively assumed that with the vm service packages too...

**Comment by jonahwilliams** [comment]
In fairness, we've gotten increasingly efficient at writing new `vm_service` packages. The latest one is even autogenerated.

**Comment by natebosch** [comment]
> Are you thinking of taking on a dependency on package:stack_trace, or authoring something minimal yourself?

`StackTrace` is from `dart:core`. There is no need for new dependencies or new implementation.

**Comment by cbracken** [comment]
I was referring to this one https://pub.dev/packages/stack_trace. If we're not taking on a dep on that I'm thrilled.

**Comment by natebosch** [comment]
Yeah we'd only need that if we were exposing a `Trace` or `Chain`. No need for `StackTrace`.

**Comment by cbracken** [comment]
SGTM make it so!

**Comment by jakemac53** [comment]
Ok I will follow up here with the new API then :).

I am going to go with the approach of making `FakeTimer` a public api, with the same public fields as were used in the debug message. I will leave the `debugString` field on it as well.

`FakeAsync.pendingTimersDebugInfo` will change to `FakeAsync.pendingTimers`, and will return a `List<FakeTimer>`.

We could leave the existing `FakeAsync.pendingTimersDebugInfo` if desired for compatibility? Although I would really prefer to call it `FakeAsync.pendingTimersDebugString`...

**Comment by jakemac53** [comment]
Ok I will follow up here with the new API then :).

I am going to go with the approach of making `FakeTimer` a public api, with the same public fields as were used in the debug message. I will leave the `debugString` field on it as well.

`FakeAsync.pendingTimersDebugInfo` will change to `FakeAsync.pendingTimers`, and will return a `List<FakeTimer>`.

We could leave the existing `FakeAsync.pendingTimersDebugInfo` if desired for compatibility?

**Comment by cbracken** [comment]
> We could leave the existing `FakeAsync.pendingTimersDebugInfo` if desired for compatibility? Although I would really prefer to call it `FakeAsync.pendingTimersDebugString`...

Seems reasonable. We can add the new API in quiver and mark the current one deprecated. I did a quick sweep internally and the one non-quiver use of it is flutter_test. Easy fix.

**Comment by cbracken** [comment]
Out of sheer paranoia, goes without saying but I assume your plan is also to rename `FakeTimer.debugInfo` to `FakeTimer.debugString`?

EDIT: never mind -- you covered that :)

**Comment by jakemac53** [comment]
Ok updated, PTAL

**Review by natebosch — APPROVED** [review]

**Pull request merged**